### PR TITLE
API Add an option to disable the cache busting on the preview field

### PR DESCRIPTION
**Pull request by maxime-rainville**
People using the S3 asset store are having problem with the `vid` on the preview image because it invalidates the AWS S3 signature.

This PR adds an option to disable the cache busting on preview.

# Parent issue
* https://github.com/silverstripe/silverstripe-asset-admin/issues/1026

### CODE REVIEW

**Comment by emteknetnz** [comment]
Where did this one get to?  I vaguely remember thinking we'd deployed a fix for this?

**Comment by maxime-rainville** [comment]
Just added some API tweaks.

**Comment by emteknetnz** [comment]
Looking good, I'll give this a test in the next day or so

**Comment by emteknetnz** [comment]
Works fine, though seems like the surrounding `<a>` tag should also have `?vid` removed when bust_cache == false ?

image

**Comment by apuckey** [comment]
> Works fine, though seems like the surrounding `<a>` tag should also have `?vid` removed when bust_cache == false ?
> 
> image

should that not also be an s3 href link? :D

**Comment by emteknetnz** [comment]
Yeah that too.  It seems like it should be an absolute link rather than relative

**Comment by apuckey** [comment]
Bump. Any news on this being added ? :D

**Comment by maxime-rainville** [comment]
I've fixed the link on the preview.

The URL comes from the asset store. @emteknetnz and I were using the regular AssetStore for testing, not the S3 one.

@apuckey I don't have an S3 sandbox up and running right now. If you can test this PR against a proper S3 project, that would help us confirm this indeed fixes the issue.

File: client/src/components/PreviewImageField/PreviewImageField.js [hunk 1]
```diff
@@ -27,15 +27,16 @@ class PreviewImageField extends Component {
     this.handleRemoveErroredUpload = this.handleRemoveErroredUpload.bind(this);
     this.canFileUpload = this.canFileUpload.bind(this);
     this.updateFormData = this.updateFormData.bind(this);
+    this.cacheBustUrl = this.cacheBustUrl.bind(this);
   }
 
-  componentWillReceiveProps(nextProps) {
+  componentDidUpdate(prevProps) {
```
Comment (maxime-rainville) [inline client/src/components/PreviewImageField/PreviewImageField.js:33]: This change is just here to squash a react deprecation warning.

**Comment by apuckey** [comment]
> I've fixed the link on the preview.
> 
> The URL comes from the asset store. @emteknetnz and I were using the regular AssetStore for testing, not the S3 one.
> 
> @apuckey I don't have an S3 sandbox up and running right now. If you can test this PR against a proper S3 project, that would help us confirm this indeed fixes the issue.

Sorry I was busy doing other things, I do have a test site operational in which I can test this out, I will let you know asap.

:)

**Comment by apuckey** [comment]
> > I've fixed the link on the preview.
> > The URL comes from the asset store. @emteknetnz and I were using the regular AssetStore for testing, not the S3 one.
> > @apuckey I don't have an S3 sandbox up and running right now. If you can test this PR against a proper S3 project, that would help us confirm this indeed fixes the issue.
> 
> Sorry I was busy doing other things, I do have a test site operational in which I can test this out, I will let you know asap.
> 
> :)

Having lots of trouble trying to build this with many missing dependencies, is there any documentation for setting up a dev environment to build this? 

for example:

```
vendor/silverstripe/asset-admin/client/src/boot/applyTransform.js
  1:22  error  Unable to resolve path to module 'lib/Injector'  import/no-unresolved
  1:22  error  Missing file extension for "lib/Injector"        import/extensions
```

and lots for i18n:

```
vendor/silverstripe/asset-admin/client/src/components/AssetDropzone/AssetDropzone.js
  4:18  error  Unable to resolve path to module 'i18n'            import/no-unresolved
  4:18  error  Missing file extension for "i18n"                  import/extensions
```

and various other missing deps.

cheers

**Comment by maxime-rainville** [comment]
I just rebased and rebuilt the PR. You should need to rebuild. You should be able to get the PR installed by adding these snippets to your composer file.

```json
{
    "require": {
        "silverstripe/asset-admin": "dev-pulls/1/allow-disabling-cache-busting-preview as 1.x-dev"
    },
    "repositories": [
        {
            "type": "github",
            "url": "git@github.com:open-sausages/silverstripe-asset-admin.git"
        }
    ]
}
```

If you do want to rebuild the JS code, you'll have to make sure you install the admin JS packages as well. You can follow these steps:
- Make sure you install `silverstripe/admin` from source ... if you have a `vendor/silverstripe/admin/client/src`  folder, you have installed from source.
  - If not run these commands: `rm -r vendor/silverstripe/admin; composer install --prefer-source`
- Install admin dependencies `(cd vendor/silverstripe/admin/ ; yarn install)`
- Rebuild asset-admin `(cd vendor/silverstripe/asset-admin/ ; yarn build)`

**Comment by apuckey** [comment]
Have put this on my test site and seems to work as expected after adding: 

```
SilverStripe\AssetAdmin\Controller\AssetAdmin:
  bust_cache: false
```

many thanks!

**Comment by maxime-rainville** [comment]
@emteknetnz Can you have one last look so we can get this merge for people who want to use S3 as their asset store?

**Comment by emteknetnz** [comment]
There's still a behat failure

**Comment by maxime-rainville** [comment]
Looks like an intermittent failure. Build is green now.

**Comment by thats4shaw** [comment]
Is there anything we can do to get this fix to work in 4.7.0?

The branch it's been merged into contains all the new GraphQL changes.

**Comment by emteknetnz** [comment]
So this pull-request didn't make it into the `1` branch before the `1.7` branch was created off it which was used for the `1.7.0` tag :-/

Unfortunately we're bound by semver not to backport this into `1.7` due to `private static $bust_cache = true;` being an API addition

Best I can think of here is to pin the merge commit for this in composer and alias it for the time being
```
"require": {
  "silverstripe/asset-admin": "1.x-dev#d24a281ab417c596b35233e543c15e22edbd729c as 1.7.0"
  ...
```

There's some [travis failures](https://github.com/silverstripe/silverstripe-asset-admin/runs/1472531916) on that commit, though looks like they're related to misquoting of yaml in the framework which broke with the upgrade to symfony 4, you should be able to safely ignore this

Hope this helps

**Comment by thats4shaw** [comment]
Thanks @emteknetnz - that's about where I got to but you still get `3.4.0` of the `silverstripe/graphql` module as part of `3.7.0 ` which isn't compatible if you pin it at that commit.

```
Server responded with an error.
Unknown type "FolderChildrenSortFields". Did you mean "ChildrenSortFieldType"?

Fragment "FileInterfaceFields" cannot be spread here as objects of type "readFilesConnection" can never be of type "FileInterface".

Fragment "FileFields" cannot be spread here as objects of type "readFilesConnection" can never be of type "File".

Fragment cannot be spread here as objects of type "readFilesConnection" can never be of type "Folder".

Unknown argument "sort" on field "children" of type "Folder". Did you mean "sortBy"?
```

Rolling back `silverstripe/graphql` appears to create more dependency issues elsewhere.

**Comment by thats4shaw** [comment]
I've just made a [fork](https://github.com/CORE-ED/silverstripe-asset-admin/tree/s3-cache-bust-fix) of the 1.7 branch and cherry picked the cache bust commit and rebuilt the javascript which will be fine for the moment.

**Pull request merged**